client/src/main/java/org/eclipse/hono/client/HonoConnection.java

```diff
@@ -82,6 +83,24 @@ static HonoConnection newConnection(final Vertx vertx, final ClientConfigPropert
         return new HonoConnectionImpl(vertx, clientConfigProperties);
     }
 
+    /**
+     * Creates a new connection using the default implementation.
+     * <p>
+     * <strong>Note:</strong> Instances of {@link ClientConfigProperties} are not thread safe and not immutable.
+     * They must therefore not be modified after calling this method.
+     *
+     * @param vertx The vert.x instance to use.
+     * @param connectionFactory The factory to use for creating an AMQP connection to the Hono server.
+     * @param clientConfigProperties The client properties to use.
+     * @return The newly created connection. Note that the underlying AMQP connection will not be established
+     *         until one of its <em>connect</em> methods is invoked.
+     * @throws NullPointerException if vertx or clientConfigProperties is {@code null}.
+     */
+    static HonoConnection newConnection(final Vertx vertx, final ConnectionFactory connectionFactory,
+            final ClientConfigProperties clientConfigProperties) {
+        return new HonoConnectionImpl(vertx, connectionFactory, clientConfigProperties);
+    }
+
```

Comment:
I do not really see why we need this additional constructor. In fact, I believe we once had one like this one but removed it because it was hard to make sure that the config props passed in are the same as the ones used for creating the connection factory. Also, `HonoConnectionImpl(Vertx, ClientConfigProperties)` seems to be doing the same thing as what you are doing in `AbstractAdapterConfig` or am I mistaken?

Reply:
I liked the idea of having the `adapterInstanceId` also represented in the AMQP connection `container-id` advertised in the OPEN frame. So I wanted to keep that.

Before this PR, the `adapterInstanceId` / `container-id` UUID was initialized in the `ClientConfigProperties`. So, in order to have a unique `adapterInstanceId` per `ProtocolAdapterCommandConsumerFactory`, a new `ClientConfigProperties` instance was needed. This is somethat counterintuitive, therefore I opted for handling the `container-id` in the `ConnectionFactory` instead, using a new `ConnectionFactory` instance for each `ProtocolAdapterCommandConsumerFactory` instance.

This leads to the need for accessing the `ConnectionFactory` instance here:
https://github.com/eclipse/hono/blob/f2b20a9f8d21b25205410a3cd26574bc389d1d89/service-base/src/main/java/org/eclipse/hono/service/AbstractAdapterConfig.java#L452-L459
An alternative would be adding a `HonoConnection.getConnectionFactory()` method, but I preferred the extra HonoConnection factory method.

But yes, I see your point about the config props passed in that factory method.
As a solution, I guess we could either
- remove the `ClientConfigProperties` param from `HonoConnectionImpl(Vertx, ConnectionFactory, ClientConfigProperties)` and the factory method (and adding `ConnectionFactory.getConfig`).
- introduce `HonoConnection.getConnectionFactory()`.
- or, deem all this not worth the effort and handle `adapterInstanceId` independent of the used `container-id`, keeping `adapterInstanceId` internal to the `ProtocolAdapterCommandConsumerFactory` instance.

I think I would prefer the first option here - but would also be OK with the others. WDYT?

Reply:
I am not sure if I can follow. My understanding is that we need to make sure that a call to `ProtocolAdapterCommandConsumerFactory.createCommandConsumer()` creates a receiver link with a unique source address of the form `command_internal/{unique_id}`. Now we would like to use the container-id of the underlying connection as the *unique_id*, right? Can't we then simply keep the container-id in a field of `HonoConnectionImpl` and make it available via e.g. `HonoConnection.getContainerId()` in order to use it when establishing the receiver link?

Reply:
Yes, that is also an option. I would let `HonoConnection.getContainerId()` call `connectionFactory.getContainerId()` then.
I've amended the commit here accordingly.

As I don't see a sensible default implementation of the new `HonoConnection.getContainerId()` method, I've set up the changes to be for 1.3 only (also removing unused methods).
For 1.2.2, I'll create a separate PR then, using the approach where the `ClientConfigProperties` get copied to get a separate `adapterInstanceId`.